### ISSUE TITLE
"No variable named dependency" when chaining dependency from parent (since v0.54.16)

### ISSUE THREAD

**Issue by jtackaberry** [gruntwork-io/terragrunt#3036]
**Describe the bug**

v0.54.16 has introduced a regression related to chained dependencies leading up to the parent module which in turn has a dependency.  v0.54.15 is fine, and the issue persists with v0.55.20.

Consider the following layout:

```
.
├── tfstate/
│   ├── outputs.tf
│   └── terragrunt.hcl
├── vpc/
│   ├── outputs.tf
│   ├── terragrunt.hcl
├── eks/
│   ├── main.tf
│   ├── terragrunt.hcl
└── terragrunt.hcl
```

With dependencies:

![image](https://github.com/gruntwork-io/terragrunt/assets/1280551/34f99f6c-fb24-4fee-be80-70563c5f1045)

The `tfstate` dependency is declared in the root-level terragrunt.hcl for purposes of generating `backend.tf`.  The vpc module works fine, however the eks module generates this error:

```
ERRO[0000] Error: Unknown variable

ERRO[0000]   on /home/jtackaberry/projects/terragrunt-dep-bug/terragrunt.hcl line 12, in remote_state:
ERRO[0000]   12:     path = "${dependency.tfstate.outputs.prefix}-${path_relative_to_include()}.tfstate"
ERRO[0000]
ERRO[0000] There is no variable named "dependency".
```

An error like the above is generated for each `dependency.tfstate.outputs` reference in the `remote_state` block.

We use this pattern quite a lot in our projects, where we have a standalone `tfstate` module in the repo (with no dependencies or includes) that bootstraps an S3 bucket and DynamoDB table using a custom module that provides the details as outputs.  Then `terragrunt.hcl` in the root sources this terragrunt module as a dependency to configure the backend.

**To Reproduce**

I've created a sample project at https://github.com/jtackaberry/terragrunt-dep-bug 

First, ensure you have a terragrunt version greater than or equal to v0.54.16.  Then:

```bash
# Clone the example
git clone https://github.com/jtackaberry/terragrunt-dep-bug.git

# Apply mock tfstate module -- this is fine
cd terragrunt-dep-bug/tfstate
terragrunt apply

# Apply mock vpc module -- this is fine
cd ../vpc
terragrunt apply

# Apply mock eks module -- error occurs here
cd ../eks
terragrunt apply
```

Though not reproduced above, I also believe #2974 is related, and possibly may have the same root cause.  The versions line up, certainly.

**Expected behavior**

The error message itself actually _may_ be innocuous.  The `eks` module in the example project does seem to work (insofar as `backend.tf` is correctly populated).

However the expected behavior is that of pre-0.54.16 where no error is printed.

**Versions**
- Terragrunt version: v0.54.15 (works), v0.54.16 (broken), v0.55.20 (broken)
- Terraform version: v1.5.7
- Environment details: Ubuntu 20.04, Ubuntu 22.04

**Comment by jtackaberry** [comment]
I was able to also update the example repo at https://github.com/jtackaberry/terragrunt-dep-bug to also reproduce #2974, as we're running into that one as well.

So now running `terragrunt apply` in the `eks/` directory outputs:

```
ERRO[0000] Error: Unknown variable

ERRO[0000]   on /home/jtackaberry/projects/terragrunt-dep-bug/terragrunt.hcl line 12, in remote_state:
ERRO[0000]   12:     path = "${dependency.tfstate.outputs.prefix}-${path_relative_to_include()}.tfstate"
ERRO[0000]
ERRO[0000] There is no variable named "dependency".

ERRO[0000] Could not convert include to the execution ctx to evaluate additional locals in file /home/jtackaberry/projects/terragrunt-dep-bug/vpc/terragrunt.hcl  prefix=[/home/jtackaberry/projects/terragrunt-dep-bug/vpc]
```

**Comment by CorentinRegnier** [comment]
Same issue there 
![image](https://github.com/gruntwork-io/terragrunt/assets/20766555/36f883e8-4a5d-4e64-8a28-7fb7404842c1)

**Comment by flightlesstux** [comment]
This issue is still exist on 0.58.2.

**Comment by levkohimins** [comment]
Resolved in [v0.62.2](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.62.2) release.
